Better XDG defaults for Windows

nojb opened the pull request:
This sets the following XDG defaults on Windows:

```
XDG_CONFIG_HOME --> %LOCALAPPDATA%
XDG_DATA_HOME --> %LOCALAPPDATA%
XDG_CACHE_HOME --> FOLDERID_InternetCache
```

A C call is needed to get the right path for `FOLDERID_InternetCache`. This is included in the PR, but perhaps it is considered to be too much trouble? Opinions welcome.

cc @dra27 @jonahbeckford

Fixes #5808

rgrinberg commented on otherlibs/xdg/xdg_stubs.c:
```diff
@@ -0,0 +1,51 @@
+#include <Knownfolders.h>
+#include <Shlobj.h>
+
+value xdg__get_user_cache_dir(value v_unit)
```
What about a prefix like `dune_xdg`?

nojb replied:
Done, thanks.

rgrinberg commented on otherlibs/xdg/xdg.ml:
```diff
@@ -10,28 +10,37 @@ type t =
   let home = t.home_dir in
-  make t "XDG_CACHE_HOME" (home / ".cache") (home / "Local Settings" / "Cache")
+  make t "XDG_CACHE_HOME" (home / ".cache") (fun env ->
+      match get_user_cache_dir () with
```
Isn't it easier to just make `get_user_cache_dir` return an option directly?

nojb replied:
Yes, fixed.

jonahbeckford commented:
Thanks. LGTM, notwithstanding rgrinberg's comments.

Tested the PR (really for my own notes because I haven't done a Dune PR):

```
Z:\source\dune>opam dkml init
Z:\source\dune>(& opam env --switch=$PWD) -split '\r?\n' | ForEach-Object { Invoke-Expression $_ }
Z:\source\dune>opam install bos utop
Z:\source\dune>gh pr checkout 5943
Z:\source\dune>with-dkml ocaml bootstrap.ml -j 8
Z:\source\dune>with-dkml .\dune.exe build @otherlibs/xdg/all

Z:\source\dune>utop
utop # #require "bos";;
utop # #directory "_build/default/otherlibs/xdg";;
utop # #directory "_build/default/otherlibs/xdg/.xdg.objs/byte";;
utop # #load_rec "xdg.cma";;
utop # let xdg = Xdg.create ~win32:true ~env:Bos.OS.Env.var () ;;
val xdg : Xdg.t = <abstr>
utop # Xdg.cache_dir xdg;;
- : string = "C:\\Users\\beckf\\AppData\\Local\\Microsoft\\Windows\\INetCache"
utop # Xdg.config_dir xdg;;
- : string = "C:\\Users\\beckf\\AppData\\Local"
utop # Xdg.data_dir xdg;;
- : string = "C:\\Users\\beckf\\AppData\\Local"
utop # Xdg.runtime_dir xdg;;
- : string option = None
```

emillon commented:
xref to https://github.com/ocaml-community/utop/pull/362 which needs something similar.

zapashcanon commented:
I don't remember the details because I wrote [directories](https://github.com/OCamlPro/directories) a long time ago, starting from how it's implemented in rust and java here https://github.com/dirs-dev ;  but :

> XDG_CONFIG_HOME --> %LOCALAPPDATA%
> XDG_DATA_HOME --> %LOCALAPPDATA%

Shouldn't it be appdata instead of localappdata (which is the one used for the cache) ? See [there](https://github.com/OCamlPro/directories/blob/master/src/windows/directories.ml#L34-L40=).

Moreover IIRC in some cases using `%APPDATA%` and not going [through the C function](https://github.com/OCamlPro/directories/blob/master/src/windows/bindings/win_functions_functor.ml#L33-L40=) would lead to an incorrect result.

zapashcanon commented:
> %APPDATA% is for "Roaming" data, that must travel with the user on different computers (this is a Windows-specific distinction). Data that is local to the user and the machine should go into %LOCALAPPDATA%.

Yes, that's why I believe the config and data should be and appdata and not localappdata, which is not the case in the PR, right ?

nojb commented:
> > %APPDATA% is for "Roaming" data, that must travel with the user on different computers (this is a Windows-specific distinction). Data that is local to the user and the machine should go into %LOCALAPPDATA%.
> 
> Yes, that's why I believe the config dir and the data dir should be in appdata and not localappdata, which is not the case in the PR, right ?

Sure, this may be a possibility. @jonahbeckford do you have an opinion as to whether `%APPDATA%` or `%LOCALAPPDATA%` should be used?

dra27 commented:
If one is trying to cater for roaming vs local settings properly, `config` should be loaded from _both_ places and merged. `display`, for example is a setting you'd expect to be able to roam, `cache` is one you probably don't want to. However, I think local is a sensible default, especially as it's possible to override it.

What does Dune write under `XDG_DATA_HOME`? I agree it's _likely_ that that should be local - on behalf of myself and anyone else with a roaming profile, it's irritating when thousands of files get blatted into my roaming profile for no particularly good reason!

FWIW, especially as you have the function already, I'd use `SHGetKnownFolderPath` for all of them - it's an anti-feature to be able to override the appdata directory with an environment variable (the environment variable exists to allow batch files to read the location, not for properly-written programs to use).

nojb commented:
> Could you add a CHANGES entry?

Done.

nojb commented:
The Windows CI is failing when building the bootstrap binary:

![image](https://user-images.githubusercontent.com/113560/178367318-a992fb7c-c75f-4dad-92fd-7bfcdf84a699.png)

I cannot reproduce locally so far. Any ideas @dra27?

MisterDA commented:
[This comment](https://github.com/msys2/MINGW-packages/issues/10211#issuecomment-982405733), possibly related, suggests adding `-luuid`.

emillon commented:
@nojb I rebased and added `-luuid` and this bootstraps now. I suppose we don't really need xdg support during bootstrap, do you know how to replace this with a dummy version for bootstrap?

emillon commented:
New error after bootstrap is:
```
** Cannot resolve symbols for otherlibs/xdg\libxdg_stubs.a(xdg_stubs.o):
 FOLDERID_InternetCache
 FOLDERID_LocalAppData
```

nojb commented:
> @nojb I rebased and added `-luuid` and this bootstraps now. I suppose we don't really need xdg support during bootstrap, do you know how to replace this with a dummy version for bootstrap?

Thanks for taking care of this PR @emillon! Sorry for the lag, I was away on holiday...

rgrinberg commented:
> I suppose we don't really need xdg support during bootstrap, do you know how to replace this with a dummy version for bootstrap?

I don't think this is what the PR did, but we aren't allowed to do that. We should respect xdg settings when using the dune cache for example.

emillon commented:
> the boostrapped binary is the one installed in the user's switch

oh is it? My knowledge about dune bootstrap dates from < 2.0 where we had a mini-dune. I assumed this was still the case, just with a different build strategy. Thanks for the clarification

emillon commented:
> Thanks for taking care of this PR @emillon! Sorry for the lag, I was away on holiday...

No problem!

anmonteiro commented:
I'm seeing a Windows failure when trying to build current Dune master (with this PR). 

Here's a GHA run: https://github.com/melange-re/melange/actions/runs/3256688824/jobs/5347432767

And relevant output:

```
      (cd _build/default/otherlibs/xdg && C:\npm\prefix\node_modules\esy\node_modules\esy-bash\.cygwin\bin\x86_64-w64-mingw32-gcc.exe -O2 -fno-strict-aliasing -fwrapv -mms-bitfields -g -I C:/Users/runneradmin/.esy/3__________________________________________________________/i/ocaml-4.14.0-cd77a8d9/lib/ocaml -o xdg_stubs.o -c xdg_stubs.c)
  
      xdg_stubs.c: In function 'dune_xdg__get_known_folder_path':
  
      xdg_stubs.c:35:9: warning: implicit declaration of function 'SHGetKnownFolderPath'; did you mean 'SHGetFolderPath'? [-Wimplicit-function-declaration]
  
         res = SHGetKnownFolderPath(rfid, 0, NULL, &wcp);
  
               ^~~~~~~~~~~~~~~~~~~~
  
               SHGetFolderPath
  
      xdg_stubs.c:41:38: error: 'WC_ERR_INVALID_CHARS' undeclared (first use in this function); did you mean 'MB_ERR_INVALID_CHARS'?
  
         len = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, wcp, wlen, NULL, 0, NULL, NULL);
  
                                            ^~~~~~~~~~~~~~~~~~~~
  
                                            MB_ERR_INVALID_CHARS
  
      xdg_stubs.c:41:38: note: each undeclared identifier is reported only once for each function it appears in
  
      error: command failed: "dune" "build" "-p" "xdg" "-j" "4" "@install" (exited with 1)
  
 ```

nojb commented:
> I'm seeing a Windows failure when trying to build current Dune master (with this PR).

This looks similar to #6098, which was supposedly fixed in #6109...

anmonteiro commented:
@phated could you do a better test on your end on windows with esy? I'm using 

```json
  "resolutions": {
    "@opam/dune": "ocaml/dune:dune.opam#bc0a0c3169dfd4aae72498c5fc5b1859dd6f870e"
  }```

rgrinberg commented:
Made an issue for this in #6241
